### ISSUE TITLE
ContextInternal timers should not use a close hook

### ISSUE THREAD

**Issue by vietj** [eclipse-vertx/vert.x#4412]
Internal timers should not be added as close hook. Such timers are used internally and creating them should never fail nor should they be cancelled when the the close hooks is closed.

See https://github.com/eclipse-vertx/vert.x/issues/4405

**Comment by kziomek** [comment]
Hi @vietj 
I'm attempting migration of our monolith from vertx 3.9.15 to 4.4.2 and I'm having following exception when running one of our bdd tests on 4.4.2.
Issue #4412 suggests such exception is not expected to be thrown. 
Is expected further work in Vert.x around this code?

I've learned so far this exception occurs when verticle restart is involved in the test.
Unfortunately I couldn't reproduce this issue on smaller project example to share yet.

```
java.lang.IllegalStateException
	at io.vertx.core.impl.CloseFuture.add(CloseFuture.java:56)
	at io.vertx.core.impl.ContextInternal.addCloseHook(ContextInternal.java:447)
	at io.vertx.core.impl.VertxImpl.scheduleTimeout(VertxImpl.java:559)
	at io.vertx.core.impl.VertxImpl.scheduleTimeout(VertxImpl.java:576)
	at io.vertx.core.impl.VertxImpl.setTimer(VertxImpl.java:394)
	at io.vertx.reactivex.ContextScheduler$ContextWorker$TimedAction.schedule(ContextScheduler.java:158)
	at io.vertx.reactivex.ContextScheduler$ContextWorker$TimedAction.access$100(ContextScheduler.java:141)
	at io.vertx.reactivex.ContextScheduler$ContextWorker.schedule(ContextScheduler.java:112)
	at hu.akarnokd.rxjava.interop.SchedulerV2ToSchedulerV1$WorkerV2ToWorkerV1.schedule(SchedulerV2ToSchedulerV1.java:74)
	at rx.internal.operators.OnSubscribeTimeoutTimedWithFallback$TimeoutMainSubscriber.startTimeout(OnSubscribeTimeoutTimedWithFallback.java:126)
	at rx.internal.operators.OnSubscribeTimeoutTimedWithFallback.call(OnSubscribeTimeoutTimedWithFallback.java:63)
	at rx.internal.operators.OnSubscribeTimeoutTimedWithFallback.call(OnSubscribeTimeoutTimedWithFallback.java:36)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:41)
	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:30)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:41)
	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:41)
	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:41)
	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:51)
	at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:35)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeRedo$2.call(OnSubscribeRedo.java:273)
	at hu.akarnokd.rxjava.interop.SchedulerV2ToSchedulerV1$Action0V1ToRunnable.run(SchedulerV2ToSchedulerV1.java:110)
	at io.vertx.reactivex.ContextScheduler$ContextWorker$TimedAction.run(ContextScheduler.java:190)
	at io.vertx.reactivex.ContextScheduler$ContextWorker$TimedAction.lambda$execute$1(ContextScheduler.java:175)
	at io.vertx.core.impl.ContextBase.lambda$null$0(ContextBase.java:137)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
	at io.vertx.core.impl.ContextBase.lambda$executeBlocking$1(ContextBase.java:135)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:750)
```

**Comment by vietj** [comment]
you should try to debug to see when this close hook is registered

**Comment by kziomek** [comment]
I understood CloseFuture was created on first verticle deployment with closed=false
Then on verticle undeployment CloseFuture flag changed to closed=true
After next verticle deployment CloseFuture with closed=true flag was used in one of operations and IllegalStateException was thrown.

I've managed to work around this particular issue by understanding which resources holds reference to closed CloseFuture.
When I changed the code to recreate these resources on verticle redeployment then closed CloseFuture were gone from jvm memory. 
In this particular case it was an object stored in static field holding io.vertx.reactivex.ContextScheduler -> EventLoopContext -> CloseFuture -> closed=true
